PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java

```diff
@@ -184,41 +183,71 @@ private WebApp getWebAppAccordingToConfiguration(@NotNull RunProcessHandler proc
+        processHandler.setText(GETTING_DEPLOYMENT_CREDENTIAL);
+        final PublishingProfile profile = webApp.getPublishingProfile();
+        processHandler.setText(CONNECTING_FTP);
+        final FTPClient ftp = WebAppUtils.getFtpConnection(profile);
+        int uploadCount;
+
+        if (webAppSettingModel.isDeployToRoot()) {
+            WebAppUtils.removeFtpDirectory(ftp, CONTAINER_ROOT_PATH, processHandler);
+            processHandler.setText(String.format(UPLOADING_ARTIFACT, CONTAINER_ROOT_PATH + ".war"));
+            uploadCount = uploadFileToFtp(ftp, CONTAINER_ROOT_PATH + ".war", input, processHandler);
+        } else {
+            WebAppUtils.removeFtpDirectory(ftp, WEB_APP_BASE_PATH + fileName, processHandler);
+            processHandler.setText(String.format(UPLOADING_ARTIFACT,
+                WEB_APP_BASE_PATH + webAppSettingModel.getTargetName()));
+            uploadCount = uploadFileToFtp(ftp, WEB_APP_BASE_PATH + webAppSettingModel.getTargetName(),
+                input, processHandler);
+        }
+        telemetryMap.put("artifactUploadCount", String.valueOf(uploadCount));
+
+        processHandler.setText(LOGGING_OUT);
+        ftp.logout();
+        if (ftp.isConnected()) {
+            ftp.disconnect();
+        }
+    }
+
+    private void uploadJarArtifact(@NotNull InputStream inputStream, @NotNull WebApp webApp,
+                                  @NotNull RunProcessHandler processHandler,
+                                  @NotNull Map<String, String> telemetryMap) throws IOException {
+        final String rootJarFile = "ROOT.jar";
+        processHandler.setText(String.format(UPLOADING_ARTIFACT, BASE_PATH + rootJarFile));
+        final int uploadCount = uploadFileViaZipDeploy(webApp, inputStream, rootJarFile, processHandler);
+        telemetryMap.put("artifactUploadCount", String.valueOf(uploadCount));
+    }
+
+    // Add retry logic here to avoid Kudu's socket timeout issue.
+    // More details: https://github.com/Microsoft/azure-maven-plugins/issues/339
+    private int uploadFileViaZipDeploy(@NotNull WebApp webapp, @NotNull InputStream inputStream,
+                                       @NotNull String fileName,
+                                       @NotNull RunProcessHandler processHandler) throws IOException {
+        int uploadCount = 0;
+        while (uploadCount < UPLOADING_MAX_TRY) {
+            uploadCount += 1;
+            try {
+                webapp.zipDeploy(FileUtil.zipFile(inputStream, fileName));
+                processHandler.setText(UPLOADING_SUCCESSFUL);
+                return uploadCount;
+            } catch (Exception e) {
+                processHandler.setText(
+                    String.format("Upload file via zip deploy met exception: %s, retry immediately...", e.getMessage()));
+            }
         }
+        throw new IOException(String.format("Upload failed after %d times of retry.", UPLOADING_MAX_TRY));
```

Comment:
lso take care of the error message here.

Reply:
Ang suggestion for the error message?